node/core/bitfield-signing/src/lib.rs

```diff
@@ -182,114 +178,156 @@ async fn construct_availability_bitfield(
+				// Abort jobs for deactivated leaves.
+				for leaf in &update.deactivated {
+					if let Some(handle) = running.remove(leaf) {
+						handle.abort();
+					}
+				}
+
+				for leaf in update.activated {
+					let sender = ctx.sender().clone();
+					let leaf_hash = leaf.hash;
+
+					let (fut, handle) = future::abortable(handle_active_leaves_update(
+						sender,
+						leaf,
+						keystore.clone(),
+						metrics.clone(),
+					));
+
+					running.insert(leaf_hash, handle);
+
+					ctx.spawn("bitfield-signing-job", fut.map(drop).boxed())?;
+				}
+			},
+			FromOverseer::Signal(OverseerSignal::BlockFinalized(..)) => {},
+			FromOverseer::Signal(OverseerSignal::Conclude) => return Ok(()),
+			FromOverseer::Communication { msg } => match msg {},
+		}
+	}
+}
+
+async fn handle_active_leaves_update<Sender>(
+	mut sender: Sender,
+	leaf: ActivatedLeaf,
+	keystore: SyncCryptoStorePtr,
+	metrics: Metrics,
+) -> Result<(), Error>
 where
-	Sender: overseer::BitfieldSigningSenderTrait + Unpin,
+	Sender: overseer::BitfieldSigningSenderTrait,
 {
-	type ToJob = BitfieldSigningMessage;
-	type OutgoingMessages = overseer::BitfieldSigningOutgoingMessages;
-	type Sender = Sender;
-	type Error = Error;
-	type RunArgs = SyncCryptoStorePtr;
-	type Metrics = Metrics;
-
-	const NAME: &'static str = "bitfield-signing-job";
-
-	/// Run a job for the parent block indicated
-	fn run(
-		leaf: ActivatedLeaf,
-		keystore: Self::RunArgs,
-		metrics: Self::Metrics,
-		_receiver: mpsc::Receiver<BitfieldSigningMessage>,
-		mut sender: JobSender<Sender>,
-	) -> Pin<Box<dyn Future<Output = Result<(), Self::Error>> + Send>> {
-		let metrics = metrics.clone();
-		async move {
-			if let LeafStatus::Stale = leaf.status {
-				gum::debug!(
+	if let LeafStatus::Stale = leaf.status {
+		gum::debug!(
+			target: LOG_TARGET,
+			hash = ?leaf.hash,
+			block_number =  ?leaf.number,
+			"Stale leaf - don't sign bitfields."
+		);
+		return Ok(())
+	}
+
+	let span = PerLeafSpan::new(leaf.span, "bitfield-signing");
+	let _span = span.child("delay");
+	let wait_until = Instant::now() + JOB_DELAY;
+
+	// now do all the work we can before we need to wait for the availability store
+	// if we're not a validator, we can just succeed effortlessly
+	let validator = match Validator::new(leaf.hash, keystore.clone(), &mut sender).await {
+		Ok(validator) => validator,
+		Err(util::Error::NotAValidator) => return Ok(()),
+		Err(err) => return Err(Error::Util(err)),
+	};
+
+	// wait a bit before doing anything else
+	Delay::new_at(wait_until).await?;
+
+	// this timer does not appear at the head of the function because we don't want to include
+	// JOB_DELAY each time.
+	let _timer = metrics.time_run();
+
+	drop(_span);
+	let span_availability = span.child("availability");
+
+	let bitfield = match construct_availability_bitfield(
+		leaf.hash,
+		&span_availability,
+		validator.index(),
+		&mut sender,
+	)
+	.await
+	{
+		Err(Error::Runtime(runtime_err)) => {
```

Comment:
Consider using `fatality`

Reply:
I think we should defer that refactor to a later point / separate PR.

Reply:
Created https://github.com/paritytech/polkadot/issues/5540